runtime/src/plugin.rs

```diff
@@ -252,35 +248,55 @@ impl Plugin {
 
     /// Get a function by name
     pub fn get_func(&mut self, function: impl AsRef<str>) -> Option<Func> {
-        self.instance
-            .get_func(&mut self.memory.get_mut().store_mut(), function.as_ref())
-    }
-
-    // A convenience method to set the plugin error and return a value
-    pub fn error<E>(&self, e: impl std::fmt::Debug, x: E) -> E {
-        self.store().data().set_error(e);
-        x
+        self.instance.get_func(&mut self.store, function.as_ref())
     }
 
     /// Store input in memory and initialize `Internal` pointer
-    pub fn set_input(&mut self, input: *const u8, mut len: usize) {
+    pub(crate) fn set_input(
+        &mut self,
+        input: *const u8,
+        mut len: usize,
+        tx: std::sync::mpsc::SyncSender<TimerAction>,
+    ) -> Result<(), Error> {
         if input.is_null() {
             len = 0;
         }
-        let ptr = self.memory.get();
-        let internal = self.internal_mut();
-        internal.input = input;
-        internal.input_length = len;
-        internal.memory = ptr
-    }
 
-    /// Dump memory using trace! logging
-    pub fn dump_memory(&self) {
-        self.memory().dump();
+        {
+            let store = &mut self.store as *mut _;
+            let linker = &mut self.linker as *mut _;
+            let internal = self.internal_mut();
+            internal.store = store;
+            internal.linker = linker;
+        }
+
+        let bytes = unsafe { std::slice::from_raw_parts(input, len) };
+        trace!("Input size: {}", bytes.len());
+
+        self.start_timer(&tx)?;
```

Comment:
is this related to the cancellation/timeout we intend to implement in the SDKs now?